File: Ladder_Basic/8-Sort Algorithm/148_sort-colors.java
/*
给出0，1，2的乱序，内部排成从0，到1，到2的顺序

Example
Example 1

Input : [1, 0, 1, 2]
Output : [0, 1, 1, 2]
Explanation : sort it in-place

*/

public class Solution {

    public void sortColors(int[] nums) {
        // write your code here
        int left = 0, right = nums.length-1;
        int i =0;
        while (i <= right) {
            if (nums[i]==0) {
                swap(nums, i++,left++);
            }
            else if (nums[i]==1) {
                i++;
            }
            else {
                swap(nums, i, right--);
            }
        }
    }
    
    private void swap(int[] nums, int i,int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: Python/8-Sort Algorithm/148_sort_colors.py
class Solution:
    """
    @param nums: A list of integer which is 0, 1 or 2 
    @return: nothing
    """
    def sortColors(self, nums):
        # write your code here
        left,right = 0,len(nums)-1
        i = 0
        while i<=right:
            if nums[i]==0:
                nums[i],nums[left]= nums[left],nums[i]
                i+=1
                left+=1
            elif nums[i]==1:
                i+=1
            else:
                nums[i], nums[right] = nums[right],nums[i]
                right-=1
        